File: BaseLua/src/main/java/org/anime_game_servers/lua/engine/LuaTable.java
package org.anime_game_servers.lua.engine;

import javax.annotation.Nullable;
import java.util.Set;

public interface LuaTable {
    boolean has(String key);

    @Nullable
    Object get(String key);

    @Nullable
    Object get(int key);

    int getInt(String key);

    int optInt(String key, int defaultValue);

    int getInt(int key);

    int optInt(int key, int defaultValue);

    void set(int key, int value);

    void set(String key, int value);

    double getDouble(String key);

    double optDouble(String key, double defaultValue);

    double getDouble(int key);

    double optDouble(int key, double defaultValue);

    float getFloat(String key);

    float optFloat(String key, float defaultValue);

    float getFloat(int key);

    float optFloat(int key, float defaultValue);

    void set(int key, double value);

    void set(String key, double value);

    boolean getBoolean(String key);

    boolean optBoolean(String key, boolean defaultValue);

    boolean getBoolean(int key);

    boolean optBoolean(int key, boolean defaultValue);

    void set(int key, boolean value);

    void set(String key, boolean value);

    String getString(String key);

    String optString(String key, String defaultValue);

    String getString(int key);

    String optString(int key, String defaultValue);

    void set(int key, String value);

    void set(String key, String value);

    @Nullable
    LuaTable getTable(String key);

    @Nullable
    LuaTable getTable(int key);

    void set(int key, LuaTable value);

    void set(String key, LuaTable value);

    Set<String> getKeys();

    int[] getAsIntArray();

    int getSize();

    Object getRawTable();
}

File: base/src/jvmMain/kotlin/org/anime_game_servers/lua/engine/LuaTable.kt
package org.anime_game_servers.lua.engine

interface LuaTable {
    fun has(key: String): Boolean

    fun get(key: String): Any?

    fun get(key: Int): Any?

    fun getInt(key: String): Int

    fun optInt(key: String, defaultValue: Int): Int

    fun getInt(key: Int): Int

    fun optInt(key: Int, defaultValue: Int): Int

    fun set(key: Int, value: Int)

    fun set(key: String, value: Int)

    fun getDouble(key: String): Double

    fun optDouble(key: String, defaultValue: Double): Double

    fun getDouble(key: Int): Double

    fun optDouble(key: Int, defaultValue: Double): Double

    fun getFloat(key: String): Float

    fun optFloat(key: String, defaultValue: Float): Float

    fun getFloat(key: Int): Float

    fun optFloat(key: Int, defaultValue: Float): Float

    fun set(key: Int, value: Double)

    fun set(key: String, value: Double)

    fun getBoolean(key: String): Boolean

    fun optBoolean(key: String, defaultValue: Boolean): Boolean

    fun getBoolean(key: Int): Boolean

    fun optBoolean(key: Int, defaultValue: Boolean): Boolean

    fun set(key: Int, value: Boolean)

    fun set(key: String, value: Boolean)

    fun getString(key: String): String?

    fun optString(key: String, defaultValue: String?): String?

    fun getString(key: Int): String?

    fun optString(key: Int, defaultValue: String?): String?

    fun set(key: Int, value: String)

    fun set(key: String, value: String)

    fun getTable(key: String): LuaTable?

    fun getTable(key: Int): LuaTable?

    fun set(key: Int, value: LuaTable)

    fun set(key: String, value: LuaTable)

    fun getKeys(): Set<String?>

    fun getAsIntArray(): IntArray

    fun getSize(): Int

    fun getRawTable(): Any?
}
